In [1]:
import numpy as np

* **Explanation:** This line imports the NumPy library and gives it the alias `np`. This is a convention used to make code shorter and easier to read when working with NumPy.

In [2]:
arr = np.array([1, 2, 3, 4, 5])
print(arr)  # [1 2 3 4 5]

[1 2 3 4 5]


* `np.array([1, 2, 3, 4, 5])`: Creates a NumPy array from the Python list `[1, 2, 3, 4, 5]`.
* `print(arr)`: Outputs the array `[1 2 3 4 5]` to the console.

In [3]:
arr = np.array([1, 2, 3, 4, 5])

# Indexing
print(arr[0])  # 1
print(arr[-1])  # 5

1
5


* `arr[0]`: Accesses the first element of the array (`1`).
* `arr[-1]`: Accesses the last element of the array (`5`).
* `print`: Displays these values.

In [4]:
print(arr[1:3])  # [2 3]
print(arr[:3])  # [1 2 3]
print(arr[2:])  # [3 4 5]

[2 3]
[1 2 3]
[3 4 5]


* Slicing:
    * `arr[1:3]`: Extracts elements from index 1 to 2 (`[2, 3]`).
    * `arr[:3]`: Extracts elements from the start up to index 2 (`[1, 2, 3]`).
    * `arr[2:]`: Extracts elements from index 2 to the end (`[3, 4, 5]`).

* `print`: Outputs these slices.

In [5]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])

* `arr1` **and** `arr2`: Create two separate arrays, `[1, 2, 3]` and `[4, 5, 6]`.

In [6]:
print(arr1 + arr2)  # [5 7 9]
print(arr1 * arr2)  # [ 4 10 18]

[5 7 9]
[ 4 10 18]


* **Element-wise Operations**:
    * `arr1 + arr2`: Adds corresponding elements of `arr1` and `arr2` (`[5, 7, 9]`).
    * `arr1 * arr2`: Multiplies corresponding elements of `arr1` and `arr2` (`[4, 10, 18]`).

In [7]:
print(np.add(arr1, arr2))  # [5 7 9]
print(np.multiply(arr1, arr2))  # [ 4 10 18]

[5 7 9]
[ 4 10 18]


* **NumPy Functions**:
    * `np.add(arr1, arr2)`: Adds corresponding elements, same as `arr1 + arr2`.
    * `np.multiply(arr1, arr2)`: Multiplies corresponding elements, same as `arr1 * arr2`.

In [8]:
arr1 = np.array([[1, 2], [3, 4]])
arr2 = np.array([[5, 6], [7, 8]])
print(np.dot(arr1, arr2))  # [[19 22], [43 50]]

[[19 22]
 [43 50]]


* **2D Arrays (`arr1` and `arr2`)**:
    * `arr1`: `[[1, 2], [3, 4]]`.
    * `arr2`: `[[5, 6], [7, 8]]`.
  
* `np.dot`: Computes the matrix product of `arr1` and `arr2`.
    * Result: `[[19, 22], [43, 50]]`.

In [9]:
arr = np.array([1, 2, 3, 4, 5])
print(np.mean(arr))  # 3.0
print(np.median(arr))  # 3.0
print(np.std(arr))  # 1.4142135623730951

3.0
3.0
1.4142135623730951


* **Statistical Functions**:
    * `np.mean(arr)`: Calculates the mean (average) of the array (`3.0`).
    * `np.median(arr)`: Finds the median (middle value) of the array (`3.0`).
    * `np.std(arr)`: Computes the standard deviation of the array (`1.41`).

In [10]:
arr = np.array([1, 2, 3, 4, 5, 6])

* `arr`: Creates an array `[1, 2, 3, 4, 5, 6]`.

In [11]:
print(np.reshape(arr, (2, 3))) 

[[1 2 3]
 [4 5 6]]


* `np.reshape`: Reshapes `arr` into a 2x3 array:
    * Output: `[[1, 2, 3], [4, 5, 6]]`.

In [12]:
arr = np.array([[1, 2], [3, 4]])

* `arr`: Creates a 2x2 array `[[1, 2], [3, 4]]`.

In [13]:
print(np.transpose(arr))

[[1 3]
 [2 4]]


* `np.transpose`: Transposes `arr` (swaps rows and columns):
    * Output: `[[1, 3], [2, 4]]`.

In [14]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])

* `arr1` **and** `arr2`: Re-initializes arrays `[1, 2, 3]` and `[4, 5, 6]`.

In [15]:
print(np.concatenate((arr1, arr2)))

[1 2 3 4 5 6]


* `np.concatenate`: Combines `arr1` and `arr2` into one array:
    * Output: `[1, 2, 3, 4, 5, 6]`.

In [16]:
arr = np.array([1, 2, 3, 4, 5, 6])
print(np.split(arr, 2)) 

[array([1, 2, 3]), array([4, 5, 6])]


* `np.split`: Splits `arr` into two equal parts:
    * Output: `[array([1, 2, 3]), array([4, 5, 6])]`.